lib/BuildSystem/SwiftTools.cpp

```diff
@@ -536,6 +541,50 @@ class SwiftCompilerShellCommand : public ExternalCommand {
     
     return true;
   }
+
+  std::string previousOutput;
+  ssize_t numBytes = 0;
+
+  void partialOutputAvailable(StringRef partialOutput) {
+    previousOutput.append(partialOutput);
+
+    do {
+      if (numBytes == 0) {
+        auto newLinePos = previousOutput.find("\n");
+        // Don't even have the first line yet.
+        if (newLinePos == std::string::npos) {
+          return;
+        }
+
+        StringRef numBytesString = previousOutput.substr(0, newLinePos);
+        if (StringRef(numBytesString).getAsInteger(10, numBytes)) {
+          numBytes = 0;
+          // Couldn't get number of bytes.
+          // Probably due to a compiler crash or bad output?
+          return;
+        }
+
+        // Remove the message length string.
+        previousOutput = previousOutput.substr(newLinePos+1, previousOutput.size());
```

Comment:
This is an inefficient way of processing this data, there is no need to actually move the bytes within the buffer except when we need to concatenate it for waiting for the next chunk. `StringRef` is designed to be really nice for this case, you can use it to "slide" over the underlying buffer while still making use of functions like `find` to work with the data.